### DISCUSSION TITLE
Integrate chatgpt assistants / assistants api

### DISCUSSION THREAD

**Discussion by davidfrothin** [open-webui/open-webui#3659] [General]
**Describe the solution you'd like**
I have multiple working chatgpt assistants that work well and has document search, function calling and all that. I would like to add the assistants id to open webui along with my openai api key. Then basically open webui can just behave like the UI.

**Describe alternatives you've considered**
Tried pipelines but it looks like I might have to write functions again? What other alternatives are there?

https://platform.openai.com/docs/assistants/overview

**Comment by tjbck** [comment]
Like you mentioned this would be appropriate as either a pipeline or a function, PR welcome!

**Reply by Mantonherre** [reply]
https://github.com/Mantonherre/Open-WebUI-Assistant-OpenAI.git

**Comment by PlebeiusGaragicus** [comment]
I have changed my mind - I'd really like to see support for the Assistants API.

**Reply by Mantonherre** [reply]
https://github.com/Mantonherre/Open-WebUI-Assistant-OpenAI.git

**Comment by davidfrothin** [comment]
@PlebeiusGaragicus @tjbck 

Thanks for your replies. Are there any examples of getting open webui to work like openai assistants (with function calling + knowledge)?

Also, it would have been way easier if open webui directly supported adding open ai assistant just like adding another model. That way users can configure their openai assistant and use open webui as the UI for that assistant.

**Reply by PlebeiusGaragicus** [reply]
I see your point and I'm not against it at all!  It actually would make it easy to use an OpenAI's Assistants.  I just don't want Open WebUI to become little more than a wrapper for OpenAI.  Instead I see it as a way to build your own system with similar capabilities - some coding required 😉

Personally, I'm working with LangGraph agents.  These use "flow engineering" which is a new paradigm beyond "prompt 
engineering."  YouTube has some very good tutorial walkthru's by Harrison Chase @hwchase17.

I think Pipelines are what you're looking for.  It's a separate Docker container that, once enabled, allows you to write your own AI model any way you want - function/tool calling, custom databases, external APIs, anything.  It seems VERY new, yet under-documented and still being designed, to be honest.  But it is exactly what I've been looking for.

When you make a pipeline it appears to the end user as just another model they can choose from.  Check Open WebUI's Pipelines GitHub for examples.  Look for their Integrations folder.  I feel it's what you're after.  The OpenAI way of doing it is natural language only - no code.  Pipelines seem way more powerful... some coding required.

To be honest (and I haven't played around enough to know if they allow API access to their assistants) your pipeline code can just call their Assistants API perhaps.  That may be what you're after..

**Reply by davidfrothin** [reply]
Thanks so much for your detailed response. I will try pipelines

**Reply by nomppy** [reply]
> Personally, I'm working with LangGraph agents. These use "flow engineering" which is a new paradigm beyond "prompt engineering." YouTube has some very good tutorial walkthru's by Harrison Chase @hwchase17.

I looked into this as well, and LangGraph came up as a very promising candidate. Do you have an implementation for this in Pipelines?

**Reply by PlebeiusGaragicus** [reply]
Yes.  It's not ready for release, really.  But I have a repo called "PlebChatDocker"

It's probably my 20th time starting from the beginning on this project using different tools until I finally found Open WebUI and pipelines.  But it's really coming together now!!!  Especially with the recent release of LangGraph Studio!!

My addition is that the agent can run commands.  Commands start with an `/`

So you can, for example, run `/url https://lilianweng.github.io/posts/2024-07-07-hallucination/` and it simply reply with a cleaned web scrape of that article.  Then you can converse about it in the chat.  Or you can run `/about` or `/help` to get a list of commands.

Of course, each user will be completely anonymous and pay their way using bitcoin over lightning.  That's the plan.

**Reply by PlebeiusGaragicus** [reply]
Now that I'm playing around with the pipelines repo itself... I realized my mistake.  I need to fork the pipelines repo and build off of it.  What I'm doing is extra complex.

Then, ideally I'd have a single Docker image with my pipeline (including all my graph code) as well as the user database to store token balances.

So, I'll be doing a large refactor... again.  But I'm making good progress!

**Reply by nomppy** [reply]
That's really cool! I'll definitely take a deeper look into your repo when I get back to work on this next week. 

What is the difference between your `url` command and current implementation of web rag in OI?

**Reply by PlebeiusGaragicus** [reply]
I have no idea how Open UI web rag works.  I'm underwhelmed by the file upload, though.  Until those are really revamped I'll use my own vector databases and my own web scraping directly implemented into my graphs/ agents.  I think as this project matures a lot of that will change and they'll add cool features..  For now I'd rather write the features myself as I learn how to do all this.

It's a killer feature to have the user upload their own files instead of relying on what's in my rag database.  But I'm just one dude moving very slowly so I think it will all pan out.  I also just have a very specific use case for work that I'm building these for so I'm ok not using those OI features.

**Reply by nomppy** [reply]
I see, ya. What I like about the OI implementation of file and web RAG is the `#` selector that feels intuitive to use, but I agree with you that the default embedding model is probably not very good. 

So your file upload triggers a filter in the pipelines that calls an embedding model (with possibly a web scrape before that) and stores the embeddings in a vector db hosted on another container or directly from pipelines? 

Btw how does your `/url` command work? I've only seen slash commands in OI used in the context of inserting prompt templates.

**Reply by PlebeiusGaragicus** [reply]
It's all a work in progress.  But the idea is to use the readability library to process the url into something like Safari's Reading Mode.  Then the agent will just reply with the content.  Then, the user can continue the conversation having the contents used for in-context learning.

```python
def url(request):
    split = request.user_message.split(" ")
    first_arg = split[1] if len(split) > 1 else None

    if not first_arg:
        return "⚠️ Please provide a URL.\n\n**Example:**\n```\n/article https://example.com\n```"

    if not first_arg.startswith("https://"):
        first_arg = f"https://{first_arg}"

    try:
        from readability import Document
        # first_arg = "https://tftc.io/home-and-car-insurance-providers-retreating/"
        response = requests.get( first_arg )
        doc = Document(response.content)

        article_markdown_contents = f""

        article_markdown_contents += doc.title()
        article_markdown_contents += doc.summary()
        article_markdown_contents += doc.content()
        return article_contents
```

**Reply by nomppy** [reply]
Sorry, I meant the triggering of the `url` function with `/url`, how does that work?

Also, I got back and had a chance to look at your repo. I see what you are doing to handle slash commands

> Now that I'm playing around with the pipelines repo itself... I realized my mistake. I need to fork the pipelines repo and build off of it. What I'm doing is extra complex.

Are you talking about the langserver? It does seem unnecessarily complex to host a separate server for this

**Reply by PlebeiusGaragicus** [reply]
I'm not going to use the official langserve thing that LangChain is making.  All you need to do is host your own FastAPI endpoint that invokes your graph.  So simple.

I'm also brainstorming a huge refactor.  I very much apologize for my poor coding practices and hope you didn't cry while you read my code.

I'm considering two options... since I'm requiring bitcoin micropayments for graph invocations, do I want my business logic (including my command handling logic) __inside__ or __outside__ my graph?

I'm currently doing a deep-dive into LangGraph to attempt to implement everything into a computable graph.  What I'm doing is premature optimization which should be avoided at all costs.  Since I want to make a litany of useful agents, I want to make easy the development of any one agent.  This space is growing so fast though I'm not worried.

my `/url` command essentially scrapes a webpage and replies with a cleaned version.  It doesn't invoke my graph.  But, perhaps my graph should catch `/` commands and run the needed callbacks and return, without invoking any LLM.  Same same.

**Reply by PlebeiusGaragicus** [reply]
Replying more directly - yes, my pipelines repo is necessary yet all it does is invoke my FastAPI server that invokes my graph.

I have a growing list of repos: Open WebUI docker container connects to my pipelines repo (I don't use the provided docker image as I need to customize slightly).  My pipeline is very basic and simply calls my FastAPI server which invokes my graph.

All the complexity of my `/url` command, etc, can be either contained inside my FastAPI server **or** my graph.  Recently I've concluded that I should have a very basic pipeline which calls a very basic FastAPI endpoint that then calls a very complicated agentic graph.  Also, I need to have a user database which stores account balances and other custom user parameters.

So I'm looking at managing least 3 repos.

**Reply by Mantonherre** [reply]
https://github.com/Mantonherre/Open-WebUI-Assistant-OpenAI.git

**Comment by Mantonherre** [comment]
I did this, and it works for me. Remember that it is a function, not a pipeline.

https://openwebui.com/f/mantonherre/assistant_openai/

**Reply by Mudr990** [reply]
i cant manage to get a reply, added the assistant id, but still no response, dependencies installed running via docker.

**Reply by Mantonherre** [reply]
Remember that the assistant and the OpenAI API key must be created under the same organization.

I haven't used Docker; I used Anaconda to create an environment, as well as npm and pip.

**Reply by Mudr990** [reply]
I seem to be something wrong, did the setup with anaconda aswell, same output, tried hardcoding the api key, no different result either.

**Reply by Mantonherre** [reply]
I'm not sure what it could be. Try checking the error that appears in the command prompt (CMD) when you save the code in the functions tab of Open WebUI. It’s possible that a dependency is missing.

**Reply by Mantonherre** [reply]
Try updating OpenAI with this command; the function is too recent.

pip install --upgrade openai

**Comment by mazierovictor** [comment]
@Mantonherre 

This function cannot send files to assistants.

I tested it and I can talk normally with the assistant, but when I try to send files it even sends it to the server, but it keeps processing forever

**Reply by Mantonherre** [reply]
I'm sorry, I haven't had time to investigate how to include the files yet.

**Reply by mazierovictor** [reply]
@Mantonherre 

No problem.

**Comment by DanialErfanian** [comment]
I read the discussion; what is the final decision? Are we going to have the OpenAI Assistant in WebUI as a built-in feature or not?

**Reply by mazierovictor** [reply]
@DanialErfanian 

I went for LibreChat, it has a much simpler interface and has native integration with OpenAi's Assistant API and also allows you to create agents in a very simple way.

**Reply by harrywang** [reply]
> LibreChat

i never used this. better than openwebui

**Reply by harrywang** [reply]
> I read the discussion; what is the final decision? Are we going to have the OpenAI Assistant in WebUI as a built-in feature or not?

I also want to know @tjbck thanks!

**Reply by mazierovictor** [reply]
I also want to know

**Comment by PlebeiusGaragicus** [comment]
You know - I've changed my mind back from not wanting to pursue assistants api.  But if we could integrate Anthropic's MCP's tho...

**Reply by PlebeiusGaragicus** [reply]
Sorry ya'll.  The state of the art is just changing so fast.  Opinions need to keep up.

 - Skills > MCP
 - Multi-agent frameworks > one-pass LLM
 - bash > file manipulation tools

it's crazy!  I'm hoping Open WebUI can keep up, but more and more it's just a think veneer over LLMs when we need a web interface that can handle more complex invocations.
